### PR TITLE
Transformers are coming

### PR DESCRIPTION
**Pull request by DrMegavolt**
@stipsan  This PR is for argument and reply transformers (WIP)
#339 

Argument transformers are working. 
Interface is compatible, and conversion are from original lib. 

Need general feedback. And hints to insert reply transformers 

Why pipeline as single commands have different code ?

### CODE REVIEW

**Comment by stipsan** [comment]
Hey great work!

The pipeline is likely done that way as it was a contribution from someone who didn't need transformers, just that the pipeline worked properly.

It makes sense to refactor the pipeline.js and command.js so they work together. The simplest solution would be to change command.js so that it returns a function instead of resolving the command right away.

As simple as this:
```diff
-return new Promise(resolve => resolve(emulate(...stringArgs))).asCallback(
+return () => new Promise(resolve => resolve(emulate(...stringArgs))).asCallback(
  callback
);
```

Then we can reuse the createCommand util here:
https://github.com/stipsan/ioredis-mock/blob/b03d7c4505422b2217c1dcebace53578e64131a0/src/pipeline.js#L30-L32

To this:
```js
Object.keys(commands).forEach(command => {
  const commandTransformer = createCommand(commands[command].bind(this), command);
  this[command] = (...args) => this.batch.push(commandTransformer(...args));
});
```

Then this area:
https://github.com/stipsan/ioredis-mock/blob/b03d7c4505422b2217c1dcebace53578e64131a0/src/index.js#L21-L23

Could be changed to this:
```js
Object.keys(commands).forEach(command => {
  const commandTransformer = createCommand(commands[command].bind(this), command);
  this[command] = (...args) => commandTransformer(...args)();
});
```

**Comment by DrMegavolt** [comment]
Thanks, I think I can do refactoring. In Express-Gateway we are using both ways 

and for reply transformers I'll just insert them before calling `resolve` to wrap the `emulate` result.
https://github.com/stipsan/ioredis-mock/pull/342/files#diff-2ce4a9c4c2c2bf320df2efab7881c510R23

make sense ?

**Comment by DrMegavolt** [comment]
Btw, first EG release that uses your lib is out. Thanks a lot for help
https://github.com/ExpressGateway/express-gateway/releases/tag/v1.4.0

**Comment by stipsan** [comment]
Ok great, yeah it makes sense to do it that way 😄 

Btw just noticed how you make sure the default transformers that ship with ioredis [is supported](https://github.com/stipsan/ioredis-mock/pull/342/files#diff-1fdf421c05c1140f6d71444ea2b27638R55). Nice!

And thanks for the shoutout in the release, really happy I could be of help to you guys! Collaborating on projects like this is what I love about open source 😄 It's a win-win-win 🎉

File: src/commands/hgetall.js [hunk 1]
```diff
@@ -1,3 +1,3 @@
 export function hgetall(key) {
-  return this.data.get(key) || {};
+  return this.data.get(key) || [];
```
Comment (stipsan) [inline src/commands/hgetall.js]: good catch!

**Comment by DrMegavolt** [comment]
@stipsan I took different approach, to refactor pipeline\command. 
It seems to be done. 

I don't understand why the test is not passing. I have all passed locally

File: src/commands/hgetall.js (hunk 1 above)
Comment (DrMegavolt) [reply src/commands/hgetall.js]: 👍  and this actually verifies that reply transformer is operational. hgetall reply transformer is converting array to object

**Comment by stipsan** [comment]
It looks good to me. The failing test only apply to node versions that don't implement `Map` natively.

We can kill to birds with one stone here by deleting this test:
https://github.com/stipsan/ioredis-mock/blob/b03d7c4505422b2217c1dcebace53578e64131a0/test/commands/hmset.js#L25-L39

Since hmset is no longer doing its own workaround but uses the hmset argument transformer this test is no longer needed. That will also make the build pass 😄

**Pull request merged**